amp/class-amp-service-worker.php

```diff
@@ -39,6 +39,91 @@ public function init() {
+				function () {
+					$template   = get_template();
+					$stylesheet = get_stylesheet();
+					$revision   = sprintf( '%s-v%s', $template, wp_get_theme( $template )->Version );
+					if ( $template !== $stylesheet ) {
+						$revision .= sprintf( ';%s-v%s', $stylesheet, wp_get_theme( $stylesheet )->Version );
+					}
+					$revision .= sprintf( ';user-%d', get_current_user_id() );
+
+					/*
+					 * Note that themes will need to vary the revision further by whatever is contained in the app shell.
+					 * In particular, the post_modified times of the nav menu items appearing in the header and footer should
+					 * be included in the revision.
+					 */
+					$revision .= ';' . md5(
+						wp_json_encode(
+							array(
+								'blogname'        => get_option( 'blogname' ),
+								'blogdescription' => get_option( 'blogdescription' ),
+								'site_icon_url'   => get_site_icon_url(),
+								'theme_mods'      => get_theme_mods(),
+								'version'         => PWA_VERSION,
+							)
+						)
+					);
+
+					// Include all scripts and styles in revision.
+					$enqueued_scripts = array();
+					foreach ( wp_scripts()->queue as $handle ) {
+						if ( isset( wp_scripts()->registered[ $handle ] ) ) {
+							$enqueued_scripts[ $handle ] = wp_scripts()->registered[ $handle ];
+						}
+					}
+					$enqueued_styles = array();
+					foreach ( wp_styles()->queue as $handle ) {
+						if ( isset( wp_styles()->registered[ $handle ] ) ) {
+							$enqueued_styles[ $handle ] = wp_styles()->registered[ $handle ];
+						}
+					}
+					$revision .= ';deps=' . md5( wp_json_encode( compact( 'enqueued_scripts', 'enqueued_styles' ) ) );
+
+					return array(
+						'url'      => add_query_arg( AMP_Theme_Support::APP_SHELL_COMPONENT_QUERY_VAR, 'outer', home_url( '/' ) ),
+						'revision' => $revision,
+					);
+				}
+			);
+
+			/**
+			 * Add the query var to designate that the inner app shell is being requested.
+			 *
+			 * @param array $precache_entry Precache entry.
+			 * @return array Modified precache entry.
+			 */
+			$add_inner_app_shell_component = function( $precache_entry ) {
+				$precache_entry['url'] = add_query_arg(
+					AMP_Theme_Support::APP_SHELL_COMPONENT_QUERY_VAR,
+					'inner',
+					$precache_entry['url']
+				);
+				return $precache_entry;
+			};
+			add_filter( 'wp_offline_error_precache_entry', $add_inner_app_shell_component, 100 );
+			add_filter( 'wp_server_error_precache_entry', $add_inner_app_shell_component, 100 );
+
+			// @todo There should be some query var that is used to disable navigation routing entirely so that there is no need to bypass for network.
+			// Prevent app shell from being served when requesting AMP version directly.
+			add_filter(
+				'wp_service_worker_navigation_route_blacklist_patterns',
+				function ( $blacklist_patterns ) {
+					$blacklist_patterns[] = '\?(.+&)*' . preg_quote( amp_get_slug(), '/' ) . '(=|&|$)';
+					return $blacklist_patterns;
+				}
+			);
+		}
```

Comment:
This I want to move this class over to the AMP plugin now, to be part of https://github.com/Automattic/amp-wp/pull/1519